jujugui/static/gui/src/test/test_bakery.js

```diff
@@ -136,6 +136,164 @@ describe('Bakery', function() {
+                                             failureCallback) {
+        originalCalled++;
+        assert.equal(path, 'path');
+        assert.equal(sucessCallback, success);
+        assert.equal(failureCallback, failure);
+      };
+      var originalMacaroon = macaroon.newMacaroon(
+          nacl.util.decodeUTF8('secret'), 'some id', 'a location');
+      var firstParty = nacl.box.keyPair();
+      var thirdParty = nacl.box.keyPair();
+      bakery.addThirdPartyCaveat(originalMacaroon,
+                                 '[some third party secret]', 'elsewhere',
+                                 thirdParty.publicKey, firstParty);
+      var m = macaroon.export(originalMacaroon);
+
+      var postCalled = 0;
+      var postSuccess = true;
+      var dischargeMacaroon;
+      bakery.webhandler.sendPostRequest = function(path,headers,query,d,e,f,
+                                                   g,completed) {
+        postCalled++;
+        postSuccess = (path == 'elsewhere/discharge')
+                      && (headers['Bakery-Protocol-Version'] == 1)
+                      && (headers['Content-Type'] ==
+                          'application/x-www-form-urlencoded');
+
+        var caveatObj = {};
+        try {
+          query.split('&').forEach(function(part) {
+            var item = part.split('=');
+            caveatObj[item[0]] = decodeURIComponent(item[1]);
+          });
+        } catch (ex) {
+          fail('unable to read url query params from sendPost request');
+        }
+
+        dischargeMacaroon = bakery.dischargeThirdPartyCaveat(caveatObj.id,
+          thirdParty, function(m) {});
+
+        // Call completed with a 400
+        // and Interaction Required to simulate interaction
+        completed({'target' : {
+          status: 400,
+          responseText: '{"Code": "interaction required", ' +
+                        '"Info": {"WaitURL": "/mywaiturl"}}'
+        }});
+      };
+
+      var getCalled = 0 , getSuccess = false;
+      bakery.webhandler.sendGetRequest = function(path, headers, query , d, e,
+                                                  f, completed) {
+        getCalled++;
+        getSuccess = (path == '/mywaiturl');
+
+        completed({'target' : {
+          status: 200,
+          responseText: JSON.stringify({
+            'Macaroon': macaroon.export(dischargeMacaroon)})
+        }});
+      };
+
+      bakery._requestHandlerWithInteraction(
+        'path', success, failure, {
+          target: {
+            status: 401,
+            responseText: JSON.stringify({'Info': {'Macaroon': m}}),
+            getResponseHeader: function(k) {return 'Macaroon';}
+          }
+        }
+      );
+
+      assert.equal(postCalled, 1);
+      assert.equal(postSuccess, true);
+      assert.equal(visitMethod.callCount(), 1);
+      assert.equal(getCalled, 1);
+      assert.equal(getSuccess, true);
+      assert.equal(originalCalled, 1);
+    });
```

Comment:
It looks like these two tests share a lot of setup; can you factor the shared parts out into a helper method they can both call--I think it might be clearer to read the tests that way.